env/file_system_tracer.h

```diff
@@ -265,24 +270,35 @@ class FSWritableFileTracingWrapper : public FSWritableFileWrapper {
 // FSWritableFileTracingWrapper when tracing is disabled.
 class FSWritableFilePtr {
  public:
-  FSWritableFilePtr(FSWritableFile* fs, std::shared_ptr<IOTracer> io_tracer)
-      : fs_(fs),
+  FSWritableFilePtr(std::unique_ptr<FSWritableFile>&& fs,
+                    const std::shared_ptr<IOTracer>& io_tracer)
+      : fs_(std::move(fs)),
         io_tracer_(io_tracer),
-        fs_tracer_(new FSWritableFileTracingWrapper(fs_, io_tracer_)) {}
-
-  explicit FSWritableFilePtr(FSWritableFile* fs)
-      : fs_(fs), io_tracer_(nullptr), fs_tracer_(nullptr) {}
+        fs_tracer_(new FSWritableFileTracingWrapper(fs_.get(), io_tracer_)) {}
```

Comment:
Can we make ```fs_tracer_``` of type ```FSWritableFileTracingWrapper``` and avoid the extra allocation?

Reply:
Originally, ```WritableFileWriter::Close``` checks if ```FSWritableFile``` ptr is  then return else does some work and close the file and set ```FSWritableFile``` ptr to null. (WritableFileWriter::Close can be called twice). So I kept ```FSWritableFileTracingWrapper``` pointer here so that I can reset it to nullptr in ```WritableFileWriter::Close```  and can check if its null or not by ``` if(writable_file_.get() != nullptr)```